### PR TITLE
Add test guide

### PR DESCRIPTION
**Pull request by mosabua**
## Description

Add guidelines for writing tests.

## Additional context and related issues

Converted from google doc - follow up about role on website will come after merge.

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.

However.. we should mention it somewhere quite prominently.

### CODE REVIEW

File: docs/src/main/sphinx/develop/tests.md
Comment (findepi) [inline docs/src/main/sphinx/develop/tests.md:146]: AFAIK this doesn't do anything with JUnit, does it?

If so, we either leave it out of the guidelines, or specifically call out that this works for TestNG tests only

Comment (mosabua) [reply docs/src/main/sphinx/develop/tests.md:146]: Sounds good .. I still have to update all this to be a guideline on what to do  .. vs the migration project

Comment (mosabua) [reply docs/src/main/sphinx/develop/tests.md]: Should we remove this bullet @martint ?

Comment (mosabua) [reply docs/src/main/sphinx/develop/tests.md]: Reworded as well btw.

File: docs/src/main/sphinx/develop/tests.md
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/tests.md:14]: I'd say tests should have enough coverage of execution paths without unnecessary complexity

File: docs/src/main/sphinx/develop/tests.md
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/tests.md:30]: Define product tests

File: docs/src/main/sphinx/develop/tests.md
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/tests.md:45]: Is it worth explaining what factors are limited? That we use public github actions runners and there's a limit of maximum number of runners that can be used concurrently. We could link to github docs here

File: docs/src/main/sphinx/develop/tests.md
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/tests.md:142]: Define flaky

File: docs/src/main/sphinx/develop/tests.md
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/tests.md:154]: Typo?

File: docs/src/main/sphinx/develop/tests.md
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/tests.md:164]: Is there a preferred solution?

File: docs/src/main/sphinx/develop/tests.md
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/tests.md:37]: Should we have a section about testing with SaaS, like BigQuery and Redshift? That Trino uses accounts sponsored by third parties and the credentials are not shared, so tests don't run by default in PRs, must be manually approved by maintainers and require own set of credentials to be executed locally.

File: docs/src/main/sphinx/develop/tests.md
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/tests.md:42]: I feel the "why" needs to be expanded a bit. This is mostly due to the nature of the project, where Trino supports multiple very different data sources where each connectors is extensively tested. We only run relevant tests in PRs based on impacted Maven modules, but all tests are always executed after pushing any changes to the master branch.

File: docs/src/main/sphinx/develop/tests.md
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/tests.md:173]: Is it worth mentioning how BaseConnectorTest is an exception to this? 😉

File: docs/src/main/sphinx/develop/tests.md
Comment (nineinchnick) [inline docs/src/main/sphinx/develop/tests.md:3]: There is a small paragraph about testing in the connector developer guide, maybe it should link here?

**Review by nineinchnick — APPROVED** [review]

File: docs/src/main/sphinx/develop/tests.md
Comment (hashhar) [inline docs/src/main/sphinx/develop/tests.md:301]: the problem this solves is that the test frameworks (neither JUnit not TestNG) don't allow controlling number of concurrent test methods per class.

e.g. how do you for example have a test class which mostly spends time waiting for I/O or some failure and reduce it's runtime?

File: docs/src/main/sphinx/develop/tests.md
Comment (mosabua) [reply docs/src/main/sphinx/develop/tests.md:3]: Where is that? Also .. do you think we need more in the overview page?

Comment (nineinchnick) [reply docs/src/main/sphinx/develop/tests.md:3]: I can't find it, I think I mixed something up, please disregard this comment.

File: docs/src/main/sphinx/develop/tests.md
Comment (mosabua) [inline docs/src/main/sphinx/develop/tests.md:407]: This section is still here for reference only .. I can delete any time .. imho all aspects are covered but wanted to have it here for reviewers for now

File: docs/src/main/sphinx/develop/tests.md
Comment (hashhar) [reply docs/src/main/sphinx/develop/tests.md:180]: > what if CI build just stop as soon as a test fails?

That might be useful for CI but very limiting for engineers because then you need more round-trips to find all the failures and fix them.

The question is basically: how to you ignore a test conditionally instead of marking it as failed. e.g. cannot test merge queries against JDBC connectors but can in others.

File: docs/src/main/sphinx/develop/tests.md
Comment (hashhar) [reply docs/src/main/sphinx/develop/tests.md:317]: > then you have the different classes run tests in parallel?

Still doesn't solve the problem. The concurrency of test is global in JUnit, ie. if you say 4 concurrent tests you will only have 4 no matter how many test classes you split it into.

You can only override it on a maven module basis via surefire config but that means a separate maven module (or profile) for each such "group of tests".

**Comment by mosabua** [comment]
Chatted with @electrum about this some more. All in your hands @martint ...

**Comment by mosabua** [comment]
Reminder @martint  .. I think just a quick read and merge will be okay since we plan to iterate on this anyway.

File: docs/src/main/sphinx/develop/tests.md
Comment (martint) [inline docs/src/main/sphinx/develop/tests.md:38]: We have integration tests that use TestContainers, so they are self-contained and don't rely on external infra.

File: docs/src/main/sphinx/develop/tests.md
Comment (martint) [inline docs/src/main/sphinx/develop/tests.md:27]: I would state these as "current conventions". Also, make it clear that there's a mix of legacy tests and that one should refer to these conventions when writing new tests instead of relying solely on existing code examples.

File: docs/src/main/sphinx/develop/tests.md
Comment (martint) [inline docs/src/main/sphinx/develop/tests.md:93]: focusing

File: docs/src/main/sphinx/develop/tests.md
Comment (martint) [inline docs/src/main/sphinx/develop/tests.md:101]: Qualify these with "consider using ..."

File: docs/src/main/sphinx/develop/tests.md
Comment (martint) [inline docs/src/main/sphinx/develop/tests.md:309]:
```
void test()
{
    assertThat(...).isEqualTo(...); // case corresponding to flag == true
    assertThat(...).isEqualTo(...); // case corresponding to flag == false
}
```

File: docs/src/main/sphinx/develop/tests.md
Comment (martint) [inline docs/src/main/sphinx/develop/tests.md:378]: Remove this

File: docs/src/main/sphinx/develop/tests.md
Comment (martint) [inline docs/src/main/sphinx/develop/tests.md:405]: Remove this

File: docs/src/main/sphinx/develop/tests.md
Comment (martint) [reply docs/src/main/sphinx/develop/tests.md:27]: Also, mention that everything in this doc is subject to change and evolution and is not set in stone.

**Comment by mosabua** [comment]
All updated.. imho this is good enough for merge and later refinement @martint

**Comment by mosabua** [comment]
Waiting for final review from @martint

File: docs/src/main/sphinx/develop/tests.md
Comment (martint) [inline docs/src/main/sphinx/develop/tests.md:280]: This should not be here.

**Review by martint — APPROVED** [review]

File: docs/src/main/sphinx/develop/tests.md
Comment (mosabua) [reply docs/src/main/sphinx/develop/tests.md:280]: Removed.

**Comment by mosabua** [comment]
Thank you @martint

**Pull request merged**
